<a href="https://colab.research.google.com/github/prathamesh0902/natural-language-processing/blob/master/Stats_500_hw_5_003.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Prathamesh Joshi**
STATS 500-Hw 5

**Problem 1**

In [ ]:
install.packages("faraway")
library(faraway)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘minqa’, ‘nloptr’, ‘Rcpp’, ‘RcppEigen’, ‘lme4’




In [ ]:
data(divusa)

In [ ]:
divusa[0:7,]

year divorce unemployed femlab marriage birth military
1 1920 8.0      5.2       22.70  92.0     117.9 3.2247  
2 1921 7.2     11.7       22.79  83.0     119.8 3.5614  
3 1922 6.6      6.7       22.88  79.7     111.2 2.4553  
4 1923 7.1      2.4       22.97  85.2     110.5 2.2065  
5 1924 7.2      5.0       23.06  80.3     110.9 2.2889  
6 1925 7.2      3.2       23.15  79.2     106.6 2.1735  
7 1926 7.5      1.8       23.24  78.7     102.6 2.1073

In [ ]:
#help(divusa)

In [ ]:
mdl1 <- lm(divorce~ unemployed + femlab + marriage + birth + military , divusa)
summary(mdl1)


Call:
lm(formula = divorce ~ unemployed + femlab + marriage + birth + 
    military, data = divusa)

Residuals:
    Min      1Q  Median      3Q     Max 
-3.8611 -0.8916 -0.0496  0.8650  3.8300 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  2.48784    3.39378   0.733   0.4659    
unemployed  -0.11125    0.05592  -1.989   0.0505 .  
femlab       0.38365    0.03059  12.543  < 2e-16 ***
marriage     0.11867    0.02441   4.861 6.77e-06 ***
birth       -0.12996    0.01560  -8.333 4.03e-12 ***
military    -0.02673    0.01425  -1.876   0.0647 .  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1.65 on 71 degrees of freedom
Multiple R-squared:  0.9208,	Adjusted R-squared:  0.9152 
F-statistic: 165.1 on 5 and 71 DF,  p-value: < 2.2e-16


In [ ]:
round(cor(divusa),2)

year  divorce unemployed femlab marriage birth military
year        1.00  0.88   -0.23       0.99  -0.62    -0.58  0.01   
divorce     0.88  1.00   -0.21       0.91  -0.53    -0.72  0.02   
unemployed -0.23 -0.21    1.00      -0.26  -0.27    -0.31 -0.40   
femlab      0.99  0.91   -0.26       1.00  -0.65    -0.60  0.05   
marriage   -0.62 -0.53   -0.27      -0.65   1.00     0.67  0.26   
birth      -0.58 -0.72   -0.31      -0.60   0.67     1.00  0.14   
military    0.01  0.02   -0.40       0.05   0.26     0.14  1.00

We do not see a collinearity among the variables of the model.

In [ ]:
##Condition number
X <- model.matrix(mdl1)[,-1]
e <- eigen(t(X) %*% X)
e$val

[1] 1174600.548   21261.741   16133.842    6206.181    1856.894

In [ ]:
round(sqrt(e$val[1]/e$val),3)

[1]  1.000  7.433  8.532 13.757 25.151

**Ans 1.1** The condition number is 25.151. This states that the collinearity in the model is very low.

In [ ]:
##Variance Inflation factor
round(vif(X),3)

unemployed     femlab   marriage      birth   military 
     2.253      3.613      2.865      2.585      1.250

**Ans 1.2** There is a very low collinearity between the predictors. Hence we cannot conclusively say about the significance of the predictors.

**Ans 1.3** The model does not show collinearity. The predictors for "femlab", "marriage" and "birth" are significant for the current model.

________________________
**Problem 2**

In [ ]:
data(longley)

In [ ]:
#help(longley)

In [ ]:
mdl2 <- lm(Employed ~ . , data=longley)
summary(mdl2)


Call:
lm(formula = Employed ~ ., data = longley)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.41011 -0.15767 -0.02816  0.10155  0.45539 

Coefficients:
               Estimate Std. Error t value Pr(>|t|)    
(Intercept)  -3.482e+03  8.904e+02  -3.911 0.003560 ** 
GNP.deflator  1.506e-02  8.492e-02   0.177 0.863141    
GNP          -3.582e-02  3.349e-02  -1.070 0.312681    
Unemployed   -2.020e-02  4.884e-03  -4.136 0.002535 ** 
Armed.Forces -1.033e-02  2.143e-03  -4.822 0.000944 ***
Population   -5.110e-02  2.261e-01  -0.226 0.826212    
Year          1.829e+00  4.555e-01   4.016 0.003037 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.3049 on 9 degrees of freedom
Multiple R-squared:  0.9955,	Adjusted R-squared:  0.9925 
F-statistic: 330.3 on 6 and 9 DF,  p-value: 4.984e-10


In [ ]:
##Condition number
Y <- model.matrix(mdl2)[,-1]
e2 <- eigen(t(Y) %*% Y)
e2$val

[1] 6.665299e+07 2.090730e+05 1.053550e+05 1.803976e+04 2.455730e+01
[6] 2.015117e+00

In [ ]:
round(sqrt(e2$val[1]/e2$val),3)

[1]    1.000   17.855   25.153   60.785 1647.478 5751.216

**Ans 2.1** The condition numbers are large stating that the collinearity in the model is high. Also we see that the predictors (except "Armed.Forces") are not significant.

In [ ]:
round(cor(longley),2)

GNP.deflator GNP  Unemployed Armed.Forces Population Year Employed
GNP.deflator 1.00         0.99  0.62       0.46        0.98       0.99 0.97    
GNP          0.99         1.00  0.60       0.45        0.99       1.00 0.98    
Unemployed   0.62         0.60  1.00      -0.18        0.69       0.67 0.50    
Armed.Forces 0.46         0.45 -0.18       1.00        0.36       0.42 0.46    
Population   0.98         0.99  0.69       0.36        1.00       0.99 0.96    
Year         0.99         1.00  0.67       0.42        0.99       1.00 0.97    
Employed     0.97         0.98  0.50       0.46        0.96       0.97 1.00

**Ans 2.2** We see that there is high correlation between "GNP.deflator", "GNP", "Population" and "Year

**And 2.3** The Variance Inflation factors are as follows:

In [ ]:
##Variance Inflation factor
round(vif(Y),3)

GNP.deflator          GNP   Unemployed Armed.Forces   Population         Year 
     135.532     1788.513       33.619        3.589      399.151      758.981

The VIF are large showing greater correlation between predictors.

We drop off "GNP.deflator", "Population" and "Year" variables.

In [ ]:
mdl3 <- lm(Employed ~ longley$GNP+longley$Armed.Forces+longley$Unemployed , data=longley)
summary(mdl3)


Call:
lm(formula = Employed ~ longley$GNP + longley$Armed.Forces + 
    longley$Unemployed, data = longley)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.83085 -0.22306  0.01735  0.10699  1.08090 

Coefficients:
                      Estimate Std. Error t value Pr(>|t|)    
(Intercept)          53.306461   0.716342  74.415  < 2e-16 ***
longley$GNP           0.040788   0.002207  18.485 3.49e-10 ***
longley$Armed.Forces -0.004828   0.002552  -1.892  0.08286 .  
longley$Unemployed   -0.007968   0.002134  -3.734  0.00285 ** 
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.4793 on 12 degrees of freedom
Multiple R-squared:  0.9851,	Adjusted R-squared:  0.9814 
F-statistic: 264.4 on 3 and 12 DF,  p-value: 3.189e-11


The predictors now show greater significance with minor loss of prediction (R^2).